plugins/hls-rename-plugin/src/Ide/Plugin/Rename.hs

```diff
@@ -0,0 +1,318 @@
+                liftIO $ runAction
+                    "Rename.GetAnnotatedParsedModule"
+                    state
+                    (use GetAnnotatedParsedSource nfp)
+    let src = T.pack $ printA annPs
+        res = T.pack $ printA $ (fmap . fmap) updateMod annPs
+    pure $ makeDiffTextEdit src res
+
+updateExports :: [Location] -> RdrName -> HsModule GhcPs -> HsModule GhcPs
+updateExports refs newRdrName ps@HsModule{hsmodExports} =
+    ps {hsmodExports = (fmap . fmap) (map (fmap $ renameIE refs newRdrName)) hsmodExports}
+
+updateImports :: [Location] -> RdrName -> HsModule GhcPs -> HsModule GhcPs
+updateImports refs newRdrName ps@HsModule{hsmodImports} =
+    ps {hsmodImports = map (fmap renameImport) hsmodImports}
+    where
+        renameImport :: ImportDecl GhcPs -> ImportDecl GhcPs
+        renameImport importDecl@ImportDecl{ideclHiding = Just (isHiding, names)} =
+            importDecl {
+                ideclHiding =
+                    Just (isHiding, fmap (map (fmap $ renameIE refs newRdrName)) names)
+                }
+        renameImport importDecl = importDecl
+
+-- TODO: implement explicit type import/export
+renameIE :: [Location] -> RdrName -> IE GhcPs -> IE GhcPs
+renameIE refs newRdrName (IEVar xVar ieName)
+    | isRef refs ieName =
+        IEVar xVar (replaceLWrappedName ieName newRdrName)
+renameIE refs newRdrName (IEThingAbs xThing ieName)
+    | isRef refs ieName =
+        IEThingAbs xThing (replaceLWrappedName ieName newRdrName)
+renameIE refs newRdrName (IEThingAll xThingAll ieName)
+    | isRef refs ieName =
+        IEThingAll xThingAll (replaceLWrappedName ieName newRdrName)
+renameIE refs newRdrName IEThingWith{}
+    = error "not implemented explicit type import/export renames yet"
+renameIE _ _ export = export
+
+-- TODO: data constructor renames
+updateLhsDecls :: [Location] -> RdrName -> HsModule GhcPs -> HsModule GhcPs
+updateLhsDecls refs newRdrName ps@HsModule{hsmodDecls} =
+    ps {hsmodDecls = map (fmap renameLhsDecl) hsmodDecls}
+    where
+        renameLhsDecl :: HsDecl GhcPs -> HsDecl GhcPs
+        renameLhsDecl (SigD xSig (TypeSig xTySig sigNames wc)) =
+            SigD xSig $ TypeSig xTySig (map renameRdrName' sigNames) wc
+        renameLhsDecl (ValD xVal funBind@FunBind{fun_id, fun_matches = fun_matches@MG{mg_alts}})
+            = ValD xVal $ funBind {
+                    fun_id = renameRdrName' fun_id,
+                    fun_matches = fun_matches {mg_alts = fmap (map (fmap $ renameLhsMatch newRdrName)) mg_alts}
+                }
+        renameLhsDecl (TyClD xTy dataDecl@DataDecl{tcdLName, tcdDataDefn = hsDataDefn@HsDataDefn{dd_cons}})
+            = TyClD xTy $ dataDecl {
+                    tcdLName = renameRdrName' tcdLName
+                }
+        renameLhsDecl (TyClD xTy synDecl@SynDecl{tcdLName})
+            = TyClD xTy $ synDecl {
+                    tcdLName = renameRdrName' tcdLName
+                }
+        renameLhsDecl decl = decl
```

Comment:
Did you consider SYB for doing this rewrite, just like wingman does? 

That way, if newer GHC versions change the AST, this code will still work.

Reply:
I think this would be a little difficult since there are a lot of names in the decls that we do not want to rename (either because retrie is already renaming the given reference or we don't support renaming for the given type of name). 

However, I could take this approach (i.e. modifying the entire source directly and not using retrie). See my comment below.